sdk/trace/src/main/scala/org/typelevel/otel4s/sdk/trace/data/EventData.scala

```diff
@@ -0,0 +1,155 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.typelevel.otel4s.sdk
+package trace.data
+
+import cats.Hash
+import cats.Show
+import cats.syntax.show._
+import org.typelevel.otel4s.Attribute
+import org.typelevel.otel4s.semconv.trace.attributes.SemanticAttributes
+
+import java.io.PrintWriter
+import java.io.StringWriter
+import scala.concurrent.duration.FiniteDuration
+import scala.util.control.NonFatal
+
+/** Data representation of an event.
+  *
+  * @see
+  *   [[https://opentelemetry.io/docs/specs/otel/trace/api/#add-events]]
+  */
+sealed trait EventData {
+
+  /** The name of the event.
+    */
+  def name: String
+
+  /** The epoch time of the event.
+    */
+  def epoch: FiniteDuration
+
+  /** The attributes of the event.
+    */
+  def attributes: Attributes
+
+  override final def hashCode(): Int =
+    Hash[EventData].hash(this)
+
+  override final def equals(obj: Any): Boolean =
+    obj match {
+      case other: EventData => Hash[EventData].eqv(this, other)
+      case _                => false
+    }
+
+  override final def toString: String =
+    Show[EventData].show(this)
+}
+
+object EventData {
+  private final val ExceptionEventName = "exception"
+
+  /** Creates [[EventData]] with the given arguments.
+    *
+    * @param name
+    *   the name of the event
+    *
+    * @param epoch
+    *   the epoch time of the event
+    *
+    * @param attributes
+    *   the attributes to associate with the event
+    */
+  def apply(
+      name: String,
+      epoch: FiniteDuration,
+      attributes: Attributes
+  ): EventData =
+    Impl(name, epoch, attributes)
+
+  /** Creates [[EventData]] from the given exception.
+    *
+    * The name of the even will be `exception`.
+    *
+    * Exception details (name, message, and stacktrace) will be added to the
+    * attributes.
+    *
+    * @param epoch
+    *   the epoch time of the event
+    *
+    * @param exception
+    *   the exception to associate with the event
+    *
+    * @param attributes
+    *   the attributes to associate with the event
+    */
+  def fromException(
+      epoch: FiniteDuration,
+      exception: Throwable,
+      attributes: Attributes
+  ): EventData = {
+    val allAttributes = {
+      val builder = Vector.newBuilder[Attribute[_]]
+
+      builder.addOne(
+        Attribute(
+          SemanticAttributes.ExceptionType,
+          Option(exception.getClass.getName).getOrElse("")
+        )
+      )
+
+      val message = exception.getMessage
+      if (message != null) {
+        builder.addOne(Attribute(SemanticAttributes.ExceptionMessage, message))
+      }
+
+      val stringWriter = new StringWriter()
+
+      try {
+        val printWriter = new PrintWriter(stringWriter)
+        exception.printStackTrace(printWriter)
+      } catch {
+        case e if NonFatal(e) => ()
+      }
```

Comment:
What kinds of non-fatal exceptions could we get here? Sorry just nit-picking, it's not a big deal. Is the Java implementation like this?

Reply:
Valid point. AFAIK `printStackTrace` doesn't throw any